parachain/pallets/omni-bridge/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+// Copyright 2020-2024 Trust Computing GmbH.
+// This file is part of Litentry.
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::too_many_arguments)]
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		fungibles::{Balanced, Create, Mutate, Refund},
+		tokens::{Balance, Fortitude, Precision, Preservation},
+		AccountTouch,
+	},
+	PalletId,
+};
+use frame_system::{ensure_root, ensure_signed, pallet_prelude::*};
+use sp_core::H256;
+use sp_io::hashing::blake2_256;
+use sp_runtime::traits::AccountIdConversion;
+use sp_std::{prelude::*, vec};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+const MODULE_ID: PalletId = PalletId(*b"hm/ombrg");
+const DEFAULT_RELAYER_THRESHOLD: u32 = 1;
+
+// TODO: maybe using xcm Location is better
+//       but we'd need enums for all foreign types, or use GeneralIndex
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum ChainType {
+	Heima,         // this chain
+	Ethereum(u32), // with chain id
+}
+
+// We assume "chain + asset-id" can uniquely identify an asset that can
+// be bridged out/in
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct ChainAsset<AssetKind> {
+	pub chain: ChainType,
+	pub asset: AssetKind,
```

Comment:
I think we are currently using different pallets to register/create assets, it would be great to streamline creating assets for bridging via this bridge such that it's easier to create and manage assets accordingly.

Reply:
This might require another pallet - something like asset manager/registry.

I intentionally decouple it from this bridge pallet (initially I have something like `create_local_asset` and then removed it), this pallet handles bridging only but not creation/deletion of assets.

It only cares if an asset exists and can be mutated.

Reply:
So far we are good with `Balances` and `Assets` - I can imagine if we have XCM assets we might need a dedicated pallet to manage them all